bpf/proxy/service_type_change_test.go

```diff
@@ -0,0 +1,362 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy_test
+
+import (
+	"net"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/projectcalico/felix/bpf/conntrack"
+	"github.com/projectcalico/felix/bpf/mock"
+	"github.com/projectcalico/felix/bpf/nat"
+	proxy "github.com/projectcalico/felix/bpf/proxy"
+	"github.com/projectcalico/felix/ip"
+)
+
+var _ = Describe("BPF service type change", func() {
+
+	clusterIP := net.IPv4(10, 1, 0, 1)
+	extIP := net.IPv4(20, 1, 0, 1)
+	port := uint16(1234)
+	proto := v1.ProtocolTCP
+	extIPstr := "20.1.0.1"
+	npPort := int32(30333)
+	testSvc := &v1.Service{
+		TypeMeta:   typeMetaV1("Service"),
+		ObjectMeta: objectMeataV1("testService"),
+		Spec: v1.ServiceSpec{
+			ClusterIP: "10.1.0.1",
+			Type:      v1.ServiceTypeClusterIP,
+			Selector: map[string]string{
+				"app": "test",
+			},
+			Ports: []v1.ServicePort{
+				{
+					Protocol: v1.ProtocolTCP,
+					Port:     int32(port),
+				},
+			},
+		},
+	}
+
+	testSvcEps := &v1.Endpoints{
+		TypeMeta:   typeMetaV1("Endpoints"),
+		ObjectMeta: objectMeataV1("testService"),
+		Subsets: []v1.EndpointSubset{
+			{
+				Addresses: []v1.EndpointAddress{
+					{
+						IP: "10.1.2.1",
+					},
+					{
+						IP: "10.1.2.2",
+					},
+				},
+				Ports: []v1.EndpointPort{
+					{
+						Port: 1234,
+						Name: "1234",
+					},
+				},
+			},
+		},
+	}
+	k8s := fake.NewSimpleClientset(testSvc, testSvcEps)
+
+	initIP := net.IPv4(1, 1, 1, 1)
+
+	front := newMockNATMap()
+	back := newMockNATBackendMap()
+	aff := newMockAffinityMap()
+	ct := mock.NewMockMap(conntrack.MapParams)
+	p, _ := proxy.StartKubeProxy(k8s, "test-node", front, back, aff, ct, proxy.WithImmediateSync())
+	p.OnHostIPsUpdate([]net.IP{initIP})
+
+	key_clusterIP := nat.NewNATKey(clusterIP, port, proxy.ProtoV1ToIntPanic(proto))
+	key_extIP := nat.NewNATKey(extIP, port, proxy.ProtoV1ToIntPanic(proto))
+	key_extIP_with_src := nat.NewNATKeySrc(extIP, port, proxy.ProtoV1ToIntPanic(proto), ip.MustParseCIDROrIP("30.1.0.1/32").(ip.V4CIDR))
+	key_hostIP := nat.NewNATKey(initIP, uint16(npPort), proxy.ProtoV1ToIntPanic(proto))
+
+	AfterEach(func() {
+		p.Stop()
+	})
+
+	It("should update service after host ip changes", func() {
+		By("check if nat map has the cluster IP", func() {
+
+			Eventually(func() bool {
+				front.Lock()
+				defer front.Unlock()
+				_, ret1 := front.m[key_clusterIP]
+				if len(front.m) == 1 && ret1 {
+					return true
+				}
+				return false
+			}).Should(BeTrue())
+		})
+		// cluster IP -> external IP
+		By("Update the service type from ClusterIP to ExternalIP", func() {
+			setSvcTypeToExternalIP(testSvc, []string{extIPstr}, k8s)
+			Eventually(func() bool {
```

Comment:
we could wrap the testing function in something like 

```
func succeeds(conds func()) bool {
	front.Lock()
	defer front.Unlock()
        return len(interceptGomegaFailures(conds)) > 0
}       
```

and then we could use the gomega matchers to express the test like 

```
Eventually(succeeds(func() {
    Expect(front.m).To(HaveLen(2))
    Expect(front.m).To(HaveKey(key_clusterIP))
    Expect(front.m).To(HaveKey(key_extIP))
}))

Which might be a bit more readable. 

Or `ret1` & co. may become `key_clusterIP_exists` so that the final condition is clear.